### PR TITLE
Improved format and template structure

### LINKED ISSUE #666: Switching to another page while in the tutorial bugs
**Describe the bug**
Switching to another page while in the tutorial bugs

**To Reproduce**
1. Go for the first time to an assignment edit page
2. press back button in the browset
4. See visual glitch

**Expected behavior**
No tutorial

**Screenshots**
image

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Browser Firefox

### LINKED ISSUE #673: Upon loading the format editor the save icon appears
**Describe the bug**
Upon loading the format editor the save icon appears as I have changed anything (but I didnt). After the save button is pressed (once) is disappears and everything works properly again.

**To Reproduce**
1. Open Firefox
2. Go to Format Editor page
4. See FAB bottom right

**Expected behavior**
no save button as nothing is changes.

**Screenshots**
image
this is on the test.ejournal page

**Desktop (please complete the following information):**
 - OS: Ubutnu
 - Browser Firefox

**Extra nodes:**

This bug does not appear on Google Chrome

### LINKED ISSUE #684: Duplication glitch in entry templates
**Describe the bug**
When setting a template from preset-only to unlimited, the template gets duplicated.

**To Reproduce**
1. Go to assignment editor
2. Click on a template
3. Set the template to preset-only
4. Save
5. Click on the same template
6. Set the template to unlimited
7. Save
_An extra template appears_

**Expected behavior**
Only 1 template

**Screenshots**
image

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Browser Firefox & Chrome

### LINKED ISSUE #689: Template fields are in the wrong order
**Describe the bug**
Template fields have a location, this should be used in frontend to determine how to position the templates. Currently the template field can appear in a different order than specified

**To Reproduce**
1. Go to a journal
2. Click on an entry
4. See a wrong order of template fields

**Expected behavior**
A correct order

### PR DESCRIPTION
**Pull request by engelhamer**
## Description
Changed the way in which assignment formats are stored. The front- and back-end structure now correspond. Also fixed several issues regarding templates and the editor.

## Summary of changes
- Fixes #666
- Fixes #673 
- Fixes #684
- Fixes #689

### CODE REVIEW

**Comment by engelhamer** [comment]
Since this hits hard on existing formats the migration file still has to be updated for a smooth transition.

File: src/vue/src/components/timeline/TimelineNodeInfo.vue [hunk 1]
```diff
@@ -41,6 +41,7 @@ export default {
             if (this.node.entry) {
                 return this.node.entry.template.name
             } else if (this.node.template) {
+                console.log(this.node)
```
Comment (Larspolo) [inline src/vue/src/components/timeline/TimelineNodeInfo.vue]: left over debug print

File: src/vue/src/components/entry/EntryFields.vue [hunk 2]
```diff
@@ -194,6 +194,9 @@ export default {
             return this.template.field_set.filter((field, i) => (field.required || this.completeContent[i].data))
         },
     },
+    created () {
+        this.template.field_set.sort((a, b) => a.location - b.location)
+    },
```
Comment (Larspolo) [inline src/vue/src/components/entry/EntryFields.vue]: Is this the only place where it should be sorted? I see also other places where the field_set is used. For example in the template edit

File: src/vue/src/views/FormatEdit.vue [hunk 3]
```diff
@@ -50,9 +50,9 @@
                     contents.'"
                     v-intro-step="3"
                     :selected="currentNode"
-                    :nodes="nodes"
+                    :nodes="presets"
                     :edit="true"
-                    @select-node="selectNode"
+                    @select-node="(node) => currentNode = node"
```
Comment (Larspolo) [inline src/vue/src/views/FormatEdit.vue]: needs curly brackets. without, it returns the return value of the assignment

File: src/django/VLE/utils/generic_utils.py [hunk 4]
```diff
@@ -119,41 +119,60 @@ def get_published_count(entries):
...
-    is updated, the template is recreated from scratch and bound to
-    all nodes that use the previous template, if there were any.
+    - format: the format that is being edited
+    - templates: the list of temlates that should be updated or created
+
+    Since existing entries that use a template should remain untouched a copy
+    of the current template is saved in an archived state before processing any
+    changes. The distinction between existing and new templates occurs based on
+    the id: newly created templates are assigned a negative id in the format
+    editor.
```
[the remark this answers is not in the document]
Comment (engelhamer) [reply src/django/VLE/utils/generic_utils.py]: To be able to distinguish the new templates. This allows users to use them in newly created presets.

File: src/django/VLE/migrations/0022_auto_20190807_1428.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            model_name='template',
+            name='format',
+            field=models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='VLE.Format'),
+        ),
+        migrations.AddField(
+            model_name='template',
+            name='preset_only',
+            field=models.BooleanField(default=False),
+        ),
+    ]
```
[the remark this answers is not in the document]
Comment (engelhamer) [reply src/django/VLE/migrations/0022_auto_20190807_1428.py]: Yes, and migrating the old formats still needs to take place!

File: src/django/VLE/timeline.py [hunk 6]
```diff
@@ -62,12 +62,14 @@ def get_nodes(journal, user=None):
...
-    if not journal or journal.assignment.format.available_templates.count() == 0:
+    if not journal or journal.assignment.format.template_set.filter(archived=False, preset_only=False).count() == 0:
         return None
     return {
         'type': Node.ADDNODE,
         'nID': -1,
-        'templates': TemplateSerializer(journal.assignment.format.available_templates.all(), many=True).data
+        'templates': TemplateSerializer(journal.assignment.format.template_set.filter(archived=False,
+                                                                                      preset_only=False),
+                                        many=True).data
```
[the remark this answers is not in the document]
Comment (engelhamer) [reply src/django/VLE/timeline.py]: Not according to pycodestyle ;)

File: src/vue/src/views/FormatEdit.vue (hunk 3 above)
Comment (engelhamer) [reply src/vue/src/views/FormatEdit.vue]: I don't see any problems here on firefox or chrome. Could you elaborate?

Comment (Larspolo) [reply src/vue/src/views/FormatEdit.vue]: https://github.com/airbnb/javascript#arrows--implicit-return

Nothing problematic happens, its just from the styleguide
```
// bad
foo(() => bool = true);

// good
foo(() => {
  bool = true;
});
```

**Comment by mjvkeulen** [comment]
#673 is not fully fixed:

Link an existing course and assignment (not sure if existing is crucial), save icon appears and will not disappear upon saving.

**Pull request merged**